main-command/src/main/scala/sbt/internal/client/BspClient.scala

```diff
@@ -18,24 +18,45 @@ import scala.sys.process.Process
 import scala.util.control.NonFatal
 
 class BspClient private (sbtServer: Socket) {
+  private val lock = new AnyRef
+  private var terminated = false
 
-  private def transferTo(input: InputStream, output: OutputStream): Unit = {
-    val buffer = Array.ofDim[Byte](1024)
-    while (true) {
-      val size = input.read(buffer)
-      output.write(buffer, 0, size)
-      output.flush()
+  private def transferTo(input: InputStream, output: OutputStream): Thread = {
+    val thread = new Thread {
+      override def run(): Unit = {
+        val buffer = Array.ofDim[Byte](1024)
+        try {
+          while (!terminated) {
+            val size = input.read(buffer)
+            if (size == -1) {
+              terminated = true
+            } else {
+              output.write(buffer, 0, size)
+              output.flush()
+            }
+          }
+          input.close()
+          output.close()
+        } finally {
```

Comment:
If you don’t catch the exception, it will print a stacktrace. If you don’t care about that, this is fine. I’m not sure where the stacktrace would be logged with the bsp client so it’s up to you to make that call.

Reply:
The stacktrace will be printed to `System.err` which is not good because the parent process will try to interpret it as a json-RPC message...